test/e2e/nbthread_test.go

```diff
@@ -0,0 +1,128 @@
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const nbthreadRouterDeployTimeout time.Duration = 1 * time.Minute
+
+// NOTE: This test will mutate the default ingresscontroller and the
+// "cluster" ingress configuration.
+func TestRouteNbthreadIngressController(t *testing.T) {
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, defaultName, defaultAvailableConditions...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	ic, err := nbthreadGetIngressController(t, kclient, 1*time.Minute)
+	if err != nil {
+		t.Fatalf("failed to get ingress controller: %v", err)
+	}
+
+	// by default, the router deployment should have ROUTER_THREADS set to 4
+	if err := nbthreadWaitForRouterDeployment(t, kclient, nbthreadRouterDeployTimeout, ic, 4); err != nil {
+		t.Fatalf("expected router deployment to have spec.threading.count unset: %v", err)
+	}
+
+	// set spec.threading.count to a nonstandard value, 7
+	if err := nbthreadSetThreadCount(t, kclient, 1*time.Minute, 7, ic); err != nil {
+		t.Fatalf("failed to update ingresscontroller: %v", err)
+	}
+
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, defaultName, defaultAvailableConditions...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	if err := nbthreadWaitForRouterDeployment(t, kclient, nbthreadRouterDeployTimeout, ic, 7); err != nil {
+		t.Fatalf("expected router deployment to set spec.threading.count: %v", err)
+	}
+
+	// set threading.count to 0, as if it was unset
+	if err := nbthreadSetThreadCount(t, kclient, 1*time.Minute, 0, ic); err != nil {
+		t.Fatalf("failed to update ingresscontroller: %v", err)
+	}
+
+	// when set back to default, the router deployment should have ROUTER_THREADS set to 4
+	if err := nbthreadWaitForRouterDeployment(t, kclient, nbthreadRouterDeployTimeout, ic, 4); err != nil {
+		t.Fatalf("expected router deployment to have spec.threading.count unset: %v", err)
+	}
+}
+
+func nbthreadGetIngressController(t *testing.T, client client.Client, timeout time.Duration) (*operatorv1.IngressController, error) {
```

Comment:
Most of these helpers look like they would be generally useful in other tests.  Any reason not to put them in `util.go` with more general names?  At some point, we can update other test code to use the helpers.  

Also, consider calling `t.Helper()` at the start of the function, at least for this and the other general helpers, so that logging and error output references the test that called the helper.

Reply:
True, there's no reason these can't be generic helper functions. I'll move them to `util.go`.